src/main/groovy/com/palantir/gradle/gitversion/JGitDescribe.groovy

```diff
@@ -0,0 +1,123 @@
+ * for --first-parent behavior.
+ */
+class JGitDescribe implements GitDescribe {
+    private static final Logger log = LoggerFactory.getLogger(JGitDescribe.class)
+
+    private File directory
+
+    JGitDescribe(File directory) {
+        this.directory = directory
+    }
+
+    @Override
+    String describe(String prefix) {
+        Git git = Git.wrap(new FileRepository(GitCli.getRootGitDir(directory)))
+        if (!GitUtils.isRepoEmpty(git)) {
+            log.debug("Repository is empty")
+            return null
+        }
+
+        ObjectId headObjectId
+        RevCommit headCommit
+        RevWalk walk
+        try {
+            headObjectId = git.getRepository().resolve(Constants.HEAD)
+            walk = new RevWalk(git.getRepository())
+            headCommit = walk.parseCommit(headObjectId)
+        } catch (Throwable ignored) {
+            log.debug("HEAD not found")
+            return null
+        }
+
+        try {
+            List<String> revs = revList(headCommit)
+
+            Map<String, String> commitHashToTag = mapCommitsToTags(git, walk)
+
+            // Walk back commit ancestors looking for tagged one
+            for (int depth = 0; depth < revs.size(); depth++) {
+                String rev = revs.get(depth)
+                if (commitHashToTag.containsKey(rev)) {
+                    String exactTag = commitHashToTag.get(rev)
+                    // Mimics '--match=${prefix}*' flag in 'git describe --tags --exact-match'
+                    if (exactTag.startsWith(prefix)) {
+                        return depth == 0 ?
+                                exactTag : String.format("%s-%s-g%s", exactTag, depth, GitUtils.abbrevHash(revs.get(0)))
+                    }
+                }
+            }
+
+            // No tags found, so return commit hash of HEAD
+            return GitUtils.abbrevHash(headObjectId.getName())
+        } catch (Throwable t) {
+            log.debug("JGit describe failed with {}", t)
+            return null
+        }
+    }
+
+    // Mimics 'git rev-list --first-parent <commit>'
+    private List<String> revList(RevCommit commit) {
+        List<String> revs = new ArrayList<>()
+        while (commit) {
+            revs.add(commit.getName())
+            try {
+                commit = commit.getParent(0)
+            } catch (Throwable ignored) {
+                break
+            }
+        }
+        return revs
+    }
+
+    // Maps all commits returned by 'git show-ref --tags -d' to output of 'git describe --tags --exact-match <commit>'
+    private Map<String, String> mapCommitsToTags(Git git, RevWalk walk) {
+        // Maps commit hash to list of all refs pointing to given commit hash.
+        // All keys in this map should be same as commit hashes in 'git show-ref --tags -d'
+        Map<String, List<RefWithTagName>> commitHashToTags = new HashMap<>()
```

Comment:
It seems a tiny bit unnecessary to build up a whole list of `RefWithTagNames` and then use the comparator to pull out the minimum - in theory we could just do the comparison whenever we get a new tag so that we only ever hold on to one.

Not sure it's worth blocking the PR for this - I think I'd like to convert everything to java before we try to do any more refactoring!

Reply:
I really want to rewrite this in Java 8...